src/apps/service/apps/jobs/submissions/async.clj

```diff
@@ -0,0 +1,118 @@
+  (let [[first-list-path first-list] (first path-lists)
+        first-list-count             (count first-list)]
+    (when (> first-list-count (config/path-list-max-paths))
+      (max-batch-paths-exceeded (config/path-list-max-paths) first-list-path first-list-count))
+    (when-not (every? (comp (partial = first-list-count) count second) path-lists)
+      (throw+ {:type  :clojure-commons.exception/illegal-argument
+               :error "All HT Analysis Path Lists must have the same number of paths."}))
+    path-lists))
+
+(defn- get-path-list-contents
+  [user path]
+  (try+
+    (when (seq path) (data-info/get-path-list-contents user path))
+    (catch Object _
+      (log/error (:throwable &throw-context)
+                 "job submission failed: Could not get file contents of path list input.")
+      (throw+))))
+
+(defn- get-path-list-contents-map
+  [user paths]
+  (into {} (map (juxt identity (partial get-path-list-contents user)) paths)))
+
+(defn- map-slice
+  [m n]
+  (->> (map (fn [[k v]] (vector k (nth v n))) m)
+       (into {})
+       (remove-nil-values)))
+
+(defn- map-slices
+  [m]
+  (let [max-count (apply max (map (comp count val) m))]
+    (mapv (partial map-slice m) (range max-count))))
+
+
+(defn- substitute-param-values
+  [path-map config]
+  (->> (map (fn [[k v]] (vector k (get path-map v v))) config)
+       (into {})))
+
+(defn- format-submission-in-batch
+  [submission job-number path-map]
+  (let [job-suffix (str "analysis-" (inc job-number))]
+    (assoc (update-in submission [:config] (partial substitute-param-values path-map))
+      :group      (config/jex-batch-group-name)
+      :name       (str (:name submission) "-" job-suffix)
+      :output_dir (ft/path-join (:output_dir submission) job-suffix))))
+
+(defn- submit-job-in-batch
+  [apps-client user submission job-number path-map]
+  (try+
+    (submit/submit-and-register-private-job apps-client
+                                            user
+                                            (format-submission-in-batch submission job-number path-map))
+    (catch Object _
+      (log/error (:throwable &throw-context)
+                 "batch job submission failed.")
+      {:status jp/failed-status})))
+
+(defn- preprocess-batch-submission
+  [submission output-dir parent-id]
+  (assoc submission
+    :output_dir           output-dir
+    :parent_id            parent-id
+    :create_output_subdir false))
+
+(defn submit-batch-jobs
+  "Asynchronously submits sub-jobs for a parent batch job,
+   where `ht-paths` is a list of paths of the HT Path List input files,
+   `submission` is the original parent submission map,
+   `output-dir` is the path to the output folder created for the parent job,
+   and `parent-id` is the parent job's ID to use for each sub-job's `parent_id` field.
+   A notification will be sent to the user on success or on failure to submit all sub-jobs."
+  [apps-client user ht-paths submission output-dir parent-id]
+  (future
```

Comment:
This is probably nitpicky, but I wonder if `future` is really the best abstraction to use here. In particular, a future is really designed for something that is going to be deref'd later -- doing computation in another thread, but not really something that autonomously does something and cleans up after itself, like this. It *works*, of course, but I think it communicates the wrong thing about the intent. Instead, I wonder if it'd be better to use either an agent or an explicit thread. For an agent, what you'd do is create an agent, then use [send-off](https://clojuredocs.org/clojure.core/send-off) or `send` to pass the function and arguments, which would then update the state of the agent. Largely, this is the same as using a future, but you could choose to not even return the agent from this function, which would make it clear the state of the agent isn't meaningful (note: futures are actually built on agents, and use the same pool as `send-off`). Using a thread directly, you'd do `(.start (Thread. (fn [] ...)))`, which doesn't even have any communication-back-to-the-parent stuff built in, so it's even more obvious. The one benefit I can see of using an agent or the future as you currently have it is that enforces the use of a thread pool, which could then be of a fixed size to limit how many huge batch jobs could be mid-submission at once, but the `send-off` pool (used by `future` as well) grows without bound from what I can read (it's using a [cached thread pool](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/Executors.html#newCachedThreadPool()) -- if you use agents and `send`, those use a fixed thread pool). If you wanted to be really fancy, you could create a new thread pool of a fixed size or caching like the send-off one, and with a naming pattern that specifies their use (e.g. the send-off/future pool uses `"clojure-agent-send-off-pool-%d"` with an atomic value for the `%d`, so maybe you'd do `"apps-submit-batch-jobs-%d"` or something), then create agents and use `send-via` to delegate the execution to that thread pool.

But it feels to me like the best expression of intent would be `(.start (Thread. (fn [] ...)))`.

Reply:
additional note, if you like the whole named thread thing but thing doing a full pool is overkill (like I suggest), you can still pass a name as the second argument to the `Thread` constructor. Could do something like name the threads `"apps-submit-batch-jobs-<uuid of parent job>"`, then the logs will be real clear about what's going on! (and note that threads can have the same name, so don't worry about potential conflicts probably, but it's still better if they are unique in my opinion)

Reply:
I think I was mostly swayed by this SO answer and went with `future` for simplicity: https://stackoverflow.com/a/8291914/1884059

I see now what you mean by the name of the thread in the logs, and I like the idea of tracking async submissions in the logs, so I think I will go ahead and switch to using `Thread` named with the parent job's ID and adding some info logging 👍